File: src/main/java/com/flexpoker/web/controller/GameManagementController.java
package com.flexpoker.web.controller;

import com.flexpoker.framework.command.CommandSender;
import com.flexpoker.game.command.commands.CreateGameCommand;
import com.flexpoker.game.command.commands.GameCommand;
import com.flexpoker.game.command.commands.JoinGameCommand;
import com.flexpoker.game.query.dto.GameInListDTO;
import com.flexpoker.game.query.dto.OpenGameForUser;
import com.flexpoker.game.query.repository.GameListRepository;
import com.flexpoker.game.query.repository.OpenGameForPlayerRepository;
import com.flexpoker.login.repository.LoginRepository;
import com.flexpoker.web.dto.incoming.CreateGameDTO;
import org.springframework.messaging.handler.annotation.MessageExceptionHandler;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.simp.annotation.SendToUser;
import org.springframework.messaging.simp.annotation.SubscribeMapping;
import org.springframework.stereotype.Controller;

import javax.inject.Inject;
import java.security.Principal;
import java.util.List;
import java.util.UUID;

@Controller
public class GameManagementController {

    private final OpenGameForPlayerRepository openGameForUserRepository;

    private final CommandSender<GameCommand> commandSender;

    private final LoginRepository loginRepository;

    private final GameListRepository gameRepository;

    @Inject
    public GameManagementController(
            OpenGameForPlayerRepository openGameForUserRepository,
            CommandSender<GameCommand> commandSender,
            LoginRepository loginRepository, GameListRepository gameRepository) {
        this.openGameForUserRepository = openGameForUserRepository;
        this.commandSender = commandSender;
        this.loginRepository = loginRepository;
        this.gameRepository = gameRepository;
    }

    @SubscribeMapping("/topic/availabletournaments")
    public List<GameInListDTO> displayAllGames() {
        return gameRepository.fetchAll();
    }

    @SubscribeMapping("/app/opengamesforuser")
    public List<OpenGameForUser> displayOpenGames(Principal principal) {
        var playerId = loginRepository.fetchAggregateIdByUsername(principal.getName());
        return openGameForUserRepository.fetchAllOpenGamesForPlayer(playerId);
    }

    @MessageMapping("/app/creategame")
    public void createGame(CreateGameDTO model, Principal principal) {
        var playerId = loginRepository.fetchAggregateIdByUsername(principal.getName());
        var command = new CreateGameCommand(model.getName(), model.getPlayers(), model.getPlayersPerTable(), playerId,
                model.getNumberOfMinutesBetweenBlindLevels(), model.getNumberOfSecondsForActionOnTimer());
        commandSender.send(command);
    }

    @MessageMapping("/app/joingame")
    public void joinGame(UUID gameId, Principal principal) {
        var playerId = loginRepository.fetchAggregateIdByUsername(principal.getName());
        var command = new JoinGameCommand(gameId, playerId);
        commandSender.send(command);
    }

    @MessageExceptionHandler
    @SendToUser("/queue/errors")
    public String handleException(Throwable exception) {
        return exception.getMessage();
    }

}

File: src/main/kotlin/com/flexpoker/web/controller/GameManagementController.kt
package com.flexpoker.web.controller

import com.flexpoker.framework.command.CommandSender
import com.flexpoker.game.command.commands.CreateGameCommand
import com.flexpoker.game.command.commands.GameCommand
import com.flexpoker.game.command.commands.JoinGameCommand
import com.flexpoker.game.query.dto.GameInListDTO
import com.flexpoker.game.query.dto.OpenGameForUser
import com.flexpoker.game.query.repository.GameListRepository
import com.flexpoker.game.query.repository.OpenGameForPlayerRepository
import com.flexpoker.login.repository.LoginRepository
import com.flexpoker.web.dto.CreateGameDTO
import org.springframework.messaging.handler.annotation.MessageExceptionHandler
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.simp.annotation.SendToUser
import org.springframework.messaging.simp.annotation.SubscribeMapping
import org.springframework.stereotype.Controller
import java.security.Principal
import java.util.UUID
import javax.inject.Inject

@Controller
class GameManagementController @Inject constructor(
    private val openGameForUserRepository: OpenGameForPlayerRepository,
    private val commandSender: CommandSender<GameCommand>,
    private val loginRepository: LoginRepository,
    private val gameRepository: GameListRepository) {

    @SubscribeMapping("/topic/availabletournaments")
    fun displayAllGames(): List<GameInListDTO> {
        return gameRepository.fetchAll()
    }

    @SubscribeMapping("/app/opengamesforuser")
    fun displayOpenGames(principal: Principal): List<OpenGameForUser>? {
        val playerId = loginRepository.fetchAggregateIdByUsername(principal.name)
        return openGameForUserRepository.fetchAllOpenGamesForPlayer(playerId)
    }

    @MessageMapping("/app/creategame")
    fun createGame(model: CreateGameDTO, principal: Principal) {
        val playerId = loginRepository.fetchAggregateIdByUsername(principal.name)
        val command = CreateGameCommand(model.name, model.players, model.playersPerTable, playerId,
            model.numberOfMinutesBetweenBlindLevels, model.numberOfSecondsForActionOnTimer)
        commandSender.send(command)
    }

    @MessageMapping("/app/joingame")
    fun joinGame(gameId: UUID, principal: Principal) {
        val playerId = loginRepository.fetchAggregateIdByUsername(principal.name)
        val command = JoinGameCommand(gameId, playerId)
        commandSender.send(command)
    }

    @MessageExceptionHandler
    @SendToUser("/queue/errors")
    fun handleException(exception: Throwable): String? {
        return exception.message
    }

}